src/coreclr/vm/comsynchronizable.cpp

```diff
@@ -795,41 +801,49 @@ void ThreadNative::InformThreadNameChange(Thread* pThread, LPCWSTR name, INT32 l
 #endif // DEBUGGING_SUPPORTED
 }
 
-FCIMPL1(FC_BOOL_RET, ThreadNative::IsThreadpoolThread, ThreadBaseObject* thread)
+// Get whether or not this is a threadpool thread.
+extern "C" BOOL QCALLTYPE ThreadNative_GetIsThreadPoolThread(QCall::ThreadHandle thread)
 {
-    FCALL_CONTRACT;
+    CONTRACTL
+    {
+        QCALL_CHECK;
+        PRECONDITION(thread != NULL);
+    }
+    CONTRACTL_END;
 
-    if (thread==NULL)
-        FCThrowRes(kNullReferenceException, W("NullReference_This"));
+    BOOL res = FALSE;
 
-    Thread *pThread = thread->GetInternal();
+    BEGIN_QCALL;
 
-    if (pThread == NULL)
-        FCThrowRes(kThreadStateException, W("ThreadState_Dead_State"));
+    if (ThreadIsDead(thread))
```

Comment:
Same here.

Reply:
Alternatively, the ThreadpoolThread flag can be duplicated on the managed side (similar to how _isDead is duplicated). The only place that reads the flag in the VM is diagnostic that we can do something with over time.

It would allow us to get rid of both:
- `ThreadNative_GetIsThreadPoolThread` - just read the bool flag on the managed thread
- `ThreadNative_SetIsThreadPoolThread` - set the bool flag on the managed thread and assert that the thread is `ThreadState.Unstarted` + pass the isThreadPool flag as an extra argument to `ThreadNative_Start` that will set the bit on the unmanaged thread for the diagnostics.